ocaml/networkd/lib/network_device_order.ml

```diff
@@ -0,0 +1,526 @@
+
+  let to_string = Xcp_pci.string_of_address
+
+  exception Parse_error of string
+
+  let of_string_exn s =
+    try Xcp_pci.address_of_string s with _ -> raise (Parse_error s)
+
+  let compare t1 t2 =
+    let open Xcp_pci in
+    match
+      ( compare t1.domain t2.domain
+      , compare t1.bus t2.bus
+      , compare t1.dev t2.dev
+      , compare t1.fn t2.fn
+      )
+    with
+    | x, _, _, _ when x <> 0 ->
+        x
+    | _, x, _, _ when x <> 0 ->
+        x
+    | _, _, x, _ when x <> 0 ->
+        x
+    | _, _, _, x ->
+        x
+end
+
+module PciaddrMap = Map.Make (Pciaddr)
+module MacaddrSet = Set.Make (Macaddr)
+module MacaddrMap = Map.Make (Macaddr)
+module IntMap = Map.Make (Int)
+module IntSet = Set.Make (Int)
+
+module ToMap = struct
+  module type Map = sig
+    type !+'a t
+
+    type key
+
+    val empty : 'a t
+
+    val update : key -> ('a option -> 'a option) -> 'a t -> 'a t
+  end
+
+  module Make (M : Map) = struct
+    exception Duplicate_key
+
+    let fail _ = raise Duplicate_key
+
+    (** [to_11_map ~by l] is a map. For each binding in the map, the key is [by v] and the value is [v] in the list. *)
+    let to_11_map ~by l =
+      List.fold_left
+        (fun acc v ->
+          let f x = Some (Option.fold ~none:v ~some:fail x) in
+          M.update (by v) f acc
+        )
+        M.empty l
+
+    (** [to_1n_map ~by m] is a map. For each binding in the map, the key is [by v] and the value is a list of [v] which are values in the [m]. *)
+    let to_1n_map ~by l =
+      List.fold_left
+        (fun acc v ->
+          let f x = Some (Option.fold ~none:[v] ~some:(List.cons v) x) in
+          M.update (by v) f acc
+        )
+        M.empty l
+  end
+end
+
+module ListToIntMap = ToMap.Make (IntMap)
+module ListToMacaddrMap = ToMap.Make (MacaddrMap)
+module ListToPciaddrMap = ToMap.Make (PciaddrMap)
+
+(** A rule is to specify a position for a network device which can be
+    identified by MAC address, PCI address, or name label. *)
+module Rule = struct
+  type index = Mac_addr of Macaddr.t | Pci_addr of Pciaddr.t | Label of string
+
+  exception Parse_error of string
+
+  type t = {position: int; index: index}
+
+  let matched ~(mac : Macaddr.t) ~(pci : Pciaddr.t) ~(label : string) t : bool =
+    match t.index with
+    | Mac_addr mac' ->
+        mac' = mac
+    | Pci_addr pci' ->
+        pci' = pci
+    | Label label' ->
+        label' = label
+
+  let parse line =
+    debug "%s: line: %s" __FUNCTION__ line ;
+    try
+      Scanf.sscanf line {|%d:%s@="%s@"|} (fun position ty value ->
+          let index =
+            match ty with
+            | "pci" ->
+                Pci_addr (Pciaddr.of_string_exn value)
+            | "mac" ->
+                Mac_addr (Macaddr.of_string_exn value)
+            | "label" ->
+                Label value
+            | _ ->
+                raise (Parse_error line)
+          in
+          {position; index}
+      )
+    with
+    | Parse_error _ as e ->
```

Comment:
Nowadays, `result` are preferred over exceptions, because the latter need to build backtraces, while the former don't.  Exceptions here look like they are used for normal flow, so I would like to see Result-base code.

Reply:
Thanks. I changed to use `Result`.